ansible-doc role arg spec support

Shrews opened the pull request:
##### SUMMARY

Change ansible-doc to support:

- Listing roles (both normal and from collections) that contain a valid role argument specification file (meta/argument_specs.yml). Text format and JSON format are supported.
- Display role arg spec documentation (text format and JSON)

##### ISSUE TYPE

- Feature Pull Request


##### COMPONENT NAME

ansible-doc

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->

### Listing Roles

```
# List roles, including a custom roles path
% ansible-doc -t role -l -r ./roles
community.general.xyz main  xyz main entry point is super awesome
community.general.xyz other xyz other entry point is super lame
sample                main  A sample role.
myuser.utils.ansible  main  Install Ansible related files.
myuser.utils.ssh      main  Install SSH configuration files.
xyz                   main  Role xyz summary that is going to be rather long so I can test line length.


# List roles, with a collection filter
% ansible-doc -t role -l myuser.utils
myuser.utils.ansible  main  Install Ansible related files.
myuser.utils.ssh      main  Install SSH configuration files.

# List roles, JSON format
% ansible-doc -t role -l --json community.general
{
    "community.general.xyz": {
        "collection": "community.general",
        "entry_points": {
            "main": "xyz main entry point is super awesome",
            "other": "xyz other entry point is super lame"
        }
    }
}
```

### Display/output of role arg spec documentation (JSON)

```
# Docs for role matching `xyz` (includes normal roles and collection roles)
% ansible-doc -t role -r ./roles --json xyz
{
    "community.general.xyz": {
        "collection": "community.general",
        "entry_points": {
            "main": {
                "short_description": "xyz main entry point is super awesome"
            },
            "other": {
                "short_description": "xyz other entry point is super lame"
            }
        },
        "path": "/Users/myuser/.ansible/collections/ansible_collections/community/general"
    },
    "xyz": {
        "collection": "",
        "entry_points": {
            "main": {
                "author": "Ansible Core (@ansible)",
                "description": [
                    "A role to demonstrate arg spec format.",
                    "Can lead to insanity."
                ],
                "options": {
                    "myopt1": {
                        "description": [
                            "First option."
                        ],
                        "required": true,
                        "type": "str"
                    },
                    "myopt2": {
                        "default": 8000,
                        "description": [
                            "Second option"
                        ],
                        "type": "int"
                    },
                    "myopt3": {
                        "choices": [
                            "choice1",
                            "choice2"
                        ],
                        "description": [
                            "Third option."
                        ],
                        "type": "str"
                    }
                },
                "short_description": "Role xyz summary that is going to be rather long so I can test line length."
            }
        },
        "path": "/Users/myuser/Devel/role_argspec/roles/xyz"
    }
}

# Docs for a FQCN role
% ansible-doc -t role --json community.general.xyz
{
    "community.general.xyz": {
        "collection": "community.general",
        "entry_points": {
            "main": {
                "short_description": "xyz main entry point is super awesome"
            },
            "other": {
                "short_description": "xyz other entry point is super lame"
            }
        },
        "path": "/Users/myuser/.ansible/collections/ansible_collections/community/general"
    }
```

### Display/output of role arg spec documentation (Text)
```
% ansible-doc -t role --playbook-dir ./ testns.testcol.testrole
> TESTNS.TESTCOL.TESTROLE    (/Users/myuser/Devel/ansible/test/integration/targets/ansible-doc/collections/ansible_collections/testns/testcol)

ENTRY POINT: main - testns.testcol.testrole short description for main entry point

        Longer description for testns.testcol.testrole main entry point.

OPTIONS (= is mandatory):

= opt1
        opt1 description

        type: str


AUTHOR: Ansible Core (@ansible)

ENTRY POINT: alternate - testns.testcol.testrole short description for alternate entry point

        Longer description for testns.testcol.testrole alternate entry point.

OPTIONS (= is mandatory):

= altopt1
        altopt1 description

        type: int


AUTHOR: Ansible Core (@ansible)


# With an entry point filter
% ansible-doc -t role --playbook-dir . testns.testcol.testrole -e alternate
> TESTNS.TESTCOL.TESTROLE    (/Users/myuser/Devel/ansible/test/integration/targets/ansible-doc/collections/ansible_collections/testns/testcol)

ENTRY POINT: alternate - testns.testcol.testrole short description for alternate entry point

        Longer description for testns.testcol.testrole alternate entry point.

OPTIONS (= is mandatory):

= altopt1
        altopt1 description

        type: int


AUTHOR: Ansible Core (@ansible)
```

Shrews committed:
Support listing roles in text and JSON

Shrews committed:
Fix json output

Shrews committed:
remove unused code

Shrews committed:
remove another unused method

Shrews committed:
Support role doc output

Shrews committed:
Support collection FQCN on role names or during listing

Shrews committed:
Add some initial tests

We need to strip the module path from the first line of the documentation output
now before comparing expected/actual output in ansible-doc/runme.sh because we now
unfrack the --playbook-dir option value.

Shrews committed:
Add role doc text output and test

Shrews committed:
Add playbook-dir roles to search path

Shrews committed:
Add entry point filter

Shrews committed:
Change tests for unfrack'd playbook_dir var

These tests were using '/tmp' for testing the setting of the playbook_dir
var. Now that we unfrack that var, MacOS will change this to '/private/tmp'
causing the tests to fail. We can choose a path that does not exist (since
unfrack does not validate existence) so that we can guarantee unfracking
will not change the value.

Shrews committed:
Cleanup fake output

s-hertel commented on lib/ansible/cli/doc.py:
```diff
@@ -63,7 +65,196 @@ class PluginNotFound(Exception):
+                summary['entry_points'][entry_point] = entry_spec.get('short_description', '')
+            result[fqcn] = summary
+
+        for role, role_path in roles:
```
If multiple role paths contain the same standalone role name build_summary overwrites one of them. First found should win to mirror the code (or maybe both should be displayed?)

Shrews replied:
Hrm, good spot. First found sounds correct to me.

Shrews replied:
Implemented first-found in latest commits.

s-hertel commented on lib/ansible/cli/doc.py:
```diff
@@ -63,7 +65,196 @@ class PluginNotFound(Exception):
+        """
+        found = set()
+        for path in role_paths:
+            if not os.path.isdir(path):
```
This seems like it would be noisy since the default role paths don't necessarily exist.

Shrews replied:
Good point. I could have swore I copied this bit from code somewhere else, but now I can't find it.

Shrews replied:
Removed in latest commits.

s-hertel commented on lib/ansible/cli/doc.py:
```diff
@@ -63,7 +65,196 @@ class PluginNotFound(Exception):
+        :returns: A set of tuples consisting of: role name, collection name, collection path
+        """
+        found = set()
+        b_colldirs = list_collection_dirs()
```
Passing the collection filter here rather than checking below will prevent importing unrelated collections in `_get_collection_name_from_path`.

Shrews replied:
Oh! Silly of me to not make use of that then.

Shrews replied:
Resolved in latest commits.

s-hertel commented on lib/ansible/cli/doc.py:
```diff
@@ -116,11 +307,19 @@ def init_parser(self):
+                                 action=opt_help.PrependListAction,
+                                 help='The path to the directory containing your roles.')
+        self.parser.add_argument("-e", "--entry-point", dest="entry_point",
+                                 help="Select the entry point for role(s).")
```
should `--list` filter by entry point if one is provided?

Shrews replied:
I don't think so. We can always add that later if needed.

s-hertel replied:
Okay. Might want to make this option mutually exclusive?

Shrews replied:
Resolved in latest commit.

s-hertel commented on lib/ansible/cli/doc.py:
```diff
@@ -172,22 +371,93 @@ def display_plugin_list(self, results):
+            # Add any 'roles' subdir in playbook dir to the roles search path
+            subdir = os.path.join(basedir, "roles")
+            if os.path.isdir(subdir):
+                roles_path = (subdir,) + roles_path
```
There are some quirks with the roles search path. If not found in the `roles` subdir, the basedir is also searched. In a collection if you have a `.` in the role name it's another directory layer - `geerlingguy.docker` in a collection should refer to `geerlingguy/docker/` not `geerlingguy.docker/`. `ansible-doc` doesn't need all of it, but it might be worth refactoring [_load_role_path](https://github.com/ansible/ansible/blob/devel/lib/ansible/playbook/role/definition.py#L137) to move the search path resolution (and maybe the role path loading?) to common code so everything stays in sync.

Shrews replied:
Hrm, I'm probably being dense here, so you may have to help me out. Not sure I follow the statement...

  `If not found in the roles subdir, the basedir is also searched.`

Is my code doing that, or code elsewhere? I don't see where that is done, and doesn't seem like correct behavior for finding roles.

The logic I've added here searches for roles in:

- DEFAULT_ROLES_PATH
- the `roles` subdir of whatever value is passed to `--playbook-dir` (which is `basedir` in this portion of the code)
- the value of `--roles-path`
- the `roles` subdir of any collections that are found.

Failing to see where the collection bit you discussed fits in, too.   :(

s-hertel replied:
FYI, the collection bit fits in here https://github.com/ansible/ansible/blob/devel/lib/ansible/playbook/role/definition.py#L159

Shrews replied:
I've added searching the playbook basedir as a last-resort in the latest commits. Thanks for seeing that undocumented behavior! I totally missed it.

After looking through the code in _load_role_path() and related bits, there is a lot of commonality, but there is also a lot of stuff happening there that ansible-doc doesn't necessarily need to do. We definitely should work to consolidate this code, but I think attempting to do so here would make this already large change much more complex. I think that should be a separate project (to be done soon, hopefully).

nitzmahone replied:
Agreed- trying to refactor/share the common bits is probably out of scope for this PR. If we end up doing the `importlib.resources` stuff though, that's probably the place to step back and take a look at where we can share a bunch of this stuff between the runtime and docs generation, and try to come up with a more coherent generalized API.

s-hertel commented on lib/ansible/cli/doc.py:
```diff
@@ -172,22 +371,93 @@ def display_plugin_list(self, results):
+            if context.CLIARGS['list_dir']:
+                # If an argument was given with --list, it is a collection filter
+                coll_filter = None
+                if len(context.CLIARGS['args']) == 1:
```
More than 1 should be handled too (it's silently ignored)

Shrews replied:
This currently matches what happens with a collection filter applied to `--list` and the other module types. I'd rather they be consistent at this point. I agree it would be nice if more than one filter was handled, but that would require modifying existing behavior not directly related to this change, so I was avoiding that (for now).

s-hertel replied:
I just meant it should either error or iterate. Right now you can pass multiple arguments and they're ignored. I didn't check what ansible-doc does for the rest of it, but I think erroring would be more clear.

Shrews replied:
We now error in latest commit.

s-hertel reviewed (commented):
This is looking good

Shrews committed:
Remove "Invalid role path" warning.

Warn when supplied more than one collection filter to --list.

Make -e/--entry-point option exclusive with --list.

Shrews committed:
Error instead of warn on multiple collection filters

Shrews committed:
Add playbook dir as last resort role search path.

If a role of the same name exists in multiple role search paths,
ensure that the first-found role is the one selected.

Shrews committed:
Fix for tests and error on invalid collection filter

Shrews committed:
Fix grep error

Shrews committed:
fix missing assignment

bcoca commented on lib/ansible/cli/doc.py:
```diff
@@ -328,6 +586,30 @@ def run(self):
+            else:
+                role_json = self._create_role_doc(context.CLIARGS['args'], roles_path,
+                                                  context.CLIARGS['entry_point'])
+                if do_json:
```
no need to do this here, we already take care of it below, just update 'docs' variable and 353/635 will display

Shrews replied:
Done

samdoran commented on lib/ansible/cli/doc.py:
```diff
@@ -71,7 +73,197 @@ class PluginNotFound(Exception):
+    Note: The methods for actual display of role data are not present here.
+    """
+
+    ROLE_ARGSPEC_FILE = 'argument_specs.yml'
```
I want to say we should support `.yaml` extension as well as `.yml`. I thought we were more flexible about this in the `galaxy` code, but it seems we only support `.yml` in many places.

I would consider that extra credit to support `.yaml` but it would be nice.

Shrews replied:
I seem to recall someone suggesting supporting only `.yml` but not sure if I'm imagining that. Will look back through my notes... but yeah, we could easily support that.

Shrews replied:
Yes, choosing only one was suggested by @nitzmahone in the spec:

```
Might want to just pick one for new features, eg, collection runtime.yml only supports .yml extension
```

samdoran commented on lib/ansible/cli/doc.py:
```diff
@@ -71,7 +73,197 @@ class PluginNotFound(Exception):
+
+        result = {}
+
+        def build_summary(role, collection, argspec):
```
Does this need to be a nested function? The overall cyclomatic complexity of this method is 8. I bet it could be reduced by removing this nested function. It would also make this method (and the function) easier to test.

Shrews replied:
TIL that "cyclomatic complexity" is a thing.  :)

Couple of questions on this:

1) Are you suggesting to just duplicate the code from `build_summary()` within each `for` loop? TBH, I would hate to do that (and was the reason for the nested function).

2) I don't see how having the nested function within `_create_role_list()` makes testing harder, but that's probably because I'm unfamiliar with cyclomatic complexity. Could you elaborate?

Shrews replied:
FYI, I'm moving the nest func out to a private class method.

samdoran replied:
> Are you suggesting to just duplicate the code from build_summary() within each for loop? 

Nope. It just seems like this nested function could be a method or a standalone function.

> I don't see how having the nested function within _create_role_list() makes testing harder,

In order to unit test `build_summary()` you would need an instance of `RoleMixin`, then call `_create_role_list()`, and then write the test in such as way as to only run `build_summary()`. It just makes it harder to get to _just_ `build_summary()` for testing, which makes it harder to throw tons of varying inputs at `build_summary()`.

samdoran replied:
> FYI, I'm moving the nest func out to a private class method.

Awesome! The complexity of `_create_role_list()` is now 4. That's what winning looks like. 😎

samdoran commented on lib/ansible/cli/doc.py:
```diff
@@ -71,7 +73,197 @@ class PluginNotFound(Exception):
+            summary['collection'] = collection
+            summary['entry_points'] = {}
+            for entry_point in argspec.keys():
+                entry_spec = argspec[entry_point] or {}
```
Under what circumstances would the `or {}`? An empty entry point inside the `argument_specs.yml` file?

```yaml
main:
  short_description: My role does amazing things

other_entry:
```

Shrews replied:
Correct. That's my attempt to keep the program from crashing if someone created an empty `argument_spec.yml` file (which is useless, but crashes aren't nice).

samdoran commented on lib/ansible/cli/doc.py:
```diff
@@ -328,6 +586,30 @@ def run(self):
+                coll_filter = None
+                if len(context.CLIARGS['args']) == 1:
+                    coll_filter = context.CLIARGS['args'][0]
+                    if not AnsibleCollectionRef.is_valid_collection_name(coll_filter):
```
This prevents listing standalone roles, such as `caddy` or `docker`. Is that intentional?

samdoran replied:
```
> ansible-doc -t role --list caddy
ERROR! Invalid collection name (must be of the form namespace.collection): caddy
```

Shrews replied:
Yes, that was intentional. Firstly, it mirrors existing `--list` behavior for other plugin types supported by `ansible-doc` where only a collection filter is supported. Secondly, it seems rather pointless to list a single role whose name we already know. Is there a use here I'm not considering?

samdoran replied:
Ok, I understand now. I hadn't ever used this with other plugin types before. It's effectively a way to list all plugin types within a collection. And now this can list all roles within a collection. Makes sense.

samdoran reviewed (approved):
I have been using this and overall like the way it works. Just a few minor questions/concerns about some of the implementations..

Shrews committed:
Remove nested func and early return

Shrews committed:
Consistent check for --list option with roles

nitzmahone reviewed (commented):
Overall LGTM. The benefit of `RoleMixin` being a mixin is unclear to me over it just being a standalone class (premature generalization?), but it's a minor nit.

This checks all the boxes for the initial asks while keeping the additions we'd need to change later for `importlib.resources` support fairly well-contained. It doesn't currently support role subdirs, but that will probably be easier to accomplish on the `Traversable` that comes back from `resources.files`, so waiting to "fix" those until then makes good sense to me.

Shrews commented:
> Overall LGTM. The benefit of `RoleMixin` being a mixin is unclear to me over it just being a standalone class (premature generalization?), but it's a minor nit.

Yep, that's exactly it.

felixfontein commented:
Would be great if the entrypoints would not be top-level, to avoid making it impossible to add top-level information later (like author, version_added, short_description, description, deprecated, ...) without resorting to ugly hacks.
